src/bnet_client.py

```diff
@@ -0,0 +1,434 @@
+
+        if header.service_id != self._RESPONSE_SERVICE_ID:
+            header.method_id = method_id
+            header.token = self._next_token()
+        else:
+            header.method_id = 0
+            header.token = method_id
+
+        header.size = body.ByteSize()
+
+        if callback:
+            self._token_callbacks[header.token] = callback
+
+        log.debug(f"protobuf sending data: service {header.service_id} ({service_name}) method {header.method_id} token {header.token}")
+
+        self.writer.write(struct.pack("!H", header.ByteSize()))
+        self.writer.write(header.SerializeToString())
+        self.writer.write(body.SerializeToString())
+        await self.writer.drain()
+
+    # @todo
+    async def _on_resource__content_handle(self, future: asyncio.Future, header, body):
+        response = content_handle_pb2.ContentHandle()
+        response.ParseFromString(body)
+        # log.debug(f"fetched content:content_handle token: {header.token} body: {response}")
+
+        future.set_result({})
+
+    async def _on_presence__query_game_account(self, future: asyncio.Future, header, body):
+        if not body:
+            log.warning("failed presence:query_response")
+            future.set_result({})
+            return
+
+        response = presence_service_pb2.QueryResponse()
+        response.ParseFromString(body)
+        # log.debug(f"fetched presence:query_response token: {header.token} body: {response}")
+
+        game_account_info = {}
+        for f in response.field:
+            if f.key.group != 2:
+                continue
+
+            if f.key.field == 1:
+                game_account_info["is_online"] = f.value.bool_value
+            elif f.key.field == 2:
+                game_account_info["is_busy"] = f.value.bool_value
+            elif f.key.field == 3:
+                game_account_info["program"] = f.value.fourcc_value  # e.g. "App", "BSAp" or "Pro" (for Overwatch)
+            elif f.key.field == 4:
+                game_account_info["last_online"] = datetime.fromtimestamp(f.value.int_value / 1000 / 1000)  # e.g. 1584194739362351 (microseconds timestamp)
+            elif f.key.field == 5:
+                continue
+            elif f.key.field == 6:
+                continue
+            elif f.key.field == 7:
+                continue
+            elif f.key.field == 8:
+                rich_presence = presence_types_pb2.RichPresence()
+                rich_presence.ParseFromString(f.value.message_value)  # e.g. "\rorP\000\025aorp\030\025"
+                game_account_info["rich_presence"] = rich_presence
+            elif f.key.field == 9:
+                continue
+            elif f.key.field == 10:
+                game_account_info["is_away"] = f.value.bool_value
+            elif f.key.field == 11:
+                continue
+
+        future.set_result(game_account_info)
+
+    async def _on_presence__query_account(self, future: asyncio.Future, header, body):
```

Comment:
What about creating `AccountInfo` class that accepts response.field on init? Probably would be easier to use typing for what the Future that have to be set than bare dictionary object. Using forward reference: https://github.com/python/typing/issues/446#issuecomment-312167911 

so then:
 `async def _on_presence__query_account(self, future: 'asyncio.Future[AccountInfo]', header, body):`

and of course for all this stuff as well for consistency. A lot of changes though.. I'm not sure if it is worth

Those signatures looks "elegant" in some way and for sure it is easier to read, but this is just my feeling. 

Another thing is those objects like `AccountInfo` can be immutable which is let say "more secure" than dealing with mutables

What do you think?

Reply:
Yes, I like your idea - it would be much cleaner. I will look into it.

Reply:
Done. Please have another look :)

Reply:
Looks good:)